### PR TITLE
make dropdowns searchable

### LINKED ISSUE #267: Replace Dropdown with Searchable Dropdown
Using this react component will make finding routes much easier while still preserving the drop down selection input: https://material-ui.com/components/autocomplete/#react-select

This should replace the from stop and to stop.

### PR DESCRIPTION
**Pull request by trxaphion**
Fixes #267 

## Proposed changes
as per the above issue the from stop and to stop selects have been replaced with react-select. I tried to emulate MUI's knick knacks as much as possible, although a few things are still unfinished. AFAIK there is:

- exit transition for menu

- limit input width

- optionally make the menu react better to resizing window horizontally while menu is open

will have a go at them sometime soon but made the PR anyway in case there's any feedback

## Link to demo, if any
https://open-transit-react-select.herokuapp.com/route/F/direction/F____O_E00

(p.s. yes I did post my username on slack)

### CODE REVIEW

**Comment by akgupta89** [comment]
I tried it on my phone and the dropdowns don't appear to work, works fine on the browser. 

Also, it'd be nice to have some cursor input indicator to reveal that this is an input box.

**Comment by trxaphion** [comment]
Thanks @akgupta89 . I should note that I don't know if the menu placement will work with iOS since I read on some old posts that their keyboard doesn't fire the resize event.

Also using the back gesture/button to close the keyboard on mobile doesn't work because the input needs to be re-focused after the resize, which of course re-opens the keyboard. I don't know a solution to this at present.

I am currently working on fixing the menu's exit transition if you want to wait for that.

File: frontend/src/actions/index.js [hunk 1]
```diff
@@ -45,8 +44,9 @@ export function fetchGraphData(params) {
   }
 }`.replace(/\s+/g, ' ');
 
-    axios.get('/api/graphql', {
-        params: { query: query, variables: JSON.stringify(params) },
+    axios
+      .get('/api/graphql', {
+        params: { query, variables: JSON.stringify(params) },
```
[the remark this answers is not in the document]
Comment (trxaphion) [reply frontend/src/actions/index.js]: sorry my bad habit of staging all files. I didn't actually make the change, I assume it was the Prettier code formatter that ran pre-commit

**Comment by hathix** [comment]
Terence makes a good point that this might look bad on small screens -- like a phone when the keyboard pops up.

Indeed, it looked pretty bad for me:

image

We agree we'd like to hide this typing input on phones. It's hard to judge what's a phone based on the screen size -- that screenshot is 2160 pixels tall (taller than my Macbook!). Judging based on the presence of a keyboard is hard and kinda flaky.

Perhaps you could listen for a `resize` event (which gets fired when a soft keyboard is opened), and in that case, use `blur` to ban typing.

Or you could do it the quick and dirty way by sniffing the user agent for Android or iOS and use [this StackOverflow answer](https://stackoverflow.com/a/22517664) to stop typing entirely.

**Comment by trxaphion** [comment]
The focus issue was due to my novice mistake of declaring the components inside the select render so they were recreated each render. 

> Noticed a weird issue on the iPhone where, after making a selection, if I tap on the select again, it keeps showing the "focus" animation (the dark line under the text input)

Closing the soft keyboard should now work but after fixing the above issue makes pressing the select afterwards blur the input instead of bringing the keyboard back up.

**Comment by hathix** [comment]
It works pretty well on my Android now -- I can see the dropdown, soft keyboard, and text input field all at once. But sometimes I get this weird problem when I click on the To field:

Screenshot_20191124-181753

**Comment by hathix** [comment]
Compare to expected behavior: 

Screenshot_20191124-182055

**Comment by trxaphion** [comment]
sorry @hathix but I couldn't reproduce your issue. The only suspect I have is your device somehow scrolled after I locked page scrolling on menu open with `document.body.style.overflow = 'hidden'`. 

I removed it from the heroku demo so see if you can reproduce it again.

**Comment by hathix** [comment]
Yes, now I'm no longer getting that problem.

**Comment by exxonvaldez** [comment]
I think I'm still seeing issues 3, 4, and 5.  But the major problems seems to be gone.  You can create new issues in github for those if you prefer.

Also, just a reminder that we merged in some css-based map effected when you hover over a dropdown menu item.  Hopefully not too hard to port over to your code, in the end it was based on some simple events.  #416

**Comment by exxonvaldez** [comment]
On narrow windows, when I start typing into the ReactSelect, the layout changes from one column to two column.  Is the react-select component missing all the logic that the normal React Select has for menu placement?  Seems like it is a lot of work to reimplement all the same layout logic.

Screen Shot 2019-12-10 at 12 04 59 AM

Screen Shot 2019-12-10 at 12 05 12 AM

**Comment by trxaphion** [comment]
No, it is the same. My menu placement only affects the dropdown menu. What you're seeing is react-select removing the placeholder/value when you type so the width is reverting to the min-width specified by the form control.

The menu going offscreen as a result is a different story though, will have to update position onkeyup maybe.

**Comment by trxaphion** [comment]
issue 4 seems to be a current react-select issue (https://github.com/JedWatson/react-select/issues/3656), and I don't see a way to explicitly set their internal focus option to work around it

aside from that I think all other issues mentioned above have been covered?

**Comment by exxonvaldez** [comment]
Looks good, thanks for putting in so many fixes.  Seems fine to wait for react-select to fix #4.

**Comment by hathix** [comment]
Let's merge this once we fix the merge conflict. Anyone want to take that on? Doesn't look too awful.

**Comment by hathix** [comment]
Wenyu, do you have the remaining merge conflicts fixed?

**Comment by hathix** [comment]
Thanks @trxaphion and @zachzwy for fixing merge conflicts!

**Comment by hathix** [comment]
Tip: if your PR is approved and just pending merge conflict fixes, you can go ahead and merge as soon as you've fixed the merge conflicts.

**Pull request merged**